### PR TITLE
React Toolbar

### PR DESCRIPTION
**Pull request by BeatC**

### CODE REVIEW

**Comment by BeatC** [comment]
During enabling react translations in toolbars I've got problem related to delaying of component rendering till appropriate language will be loaded. Because of that some legacy UI code that depends on Toolbars runs before toolbar has been actually rendered. I had conversation with @saghul about this problem and we decided to leave this branch as it is for a while. And start reactifying of toolbar dependent UI code. And then after merging that part into master continue to work on toolbars. Doing reactification this way allows us not to do this PR even larger than it is and also reactify toolbar related stuff faster and simpler (e.g. we shouldn't work on translations twice)

File: react/features/conference/components/Conference.native.js [hunk 1]
```diff
@@ -232,12 +259,41 @@ class Conference extends Component {
...
+ * @private
+ * @returns {{
+ *     _onConnect: Function,
+ *     _onDisconnect: Function,
+ *     _onToolbarVisibilityChange: Function
+ * }}
+ */
+function _mapDispatchToProps(dispatch) {
+    return {
+        _onConnect() {
```
Comment (lyubomir) [inline react/features/conference/components/Conference.native.js]: Please comply with our coding style which mandates jsdocs on everything.

File: react/features/toolbar/components/secondaryToolbarHandlers.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+            if (value && sipNumber) {
+                APP.UI.emitEvent(UIEvents.SIP_DIAL, sipNumber);
+            }
+        },
+        focus: ':input:first'
+    });
+}
+
+export default {
```
Comment (lyubomir) [inline react/features/toolbar/components/secondaryToolbarHandlers.js]: I and Yana (as far as I understand) find the separation between `primaryToolbarHandlers` and `secondaryToolbarHandlers` inappropriate. The words `primary` and `secondary` refer to UI/presentation while these are handlers i.e. (business) logic. Please remove the separation between `primaryToolbarHandlers` and `secondaryToolbarHandlers`.

File: react/features/toolbar/components/primaryToolbarHandlers.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
+/* @flow */
+
+import UIEvents from '../../../../service/UI/UIEvents';
+
+declare var APP: Object;
+declare var JitsiMeetJS: Object;
+
+export default {
+    'toolbar_button_mute': function() {
```
Comment (lyubomir) [inline react/features/toolbar/components/primaryToolbarHandlers.js]: 1. In all objects (i.e. not only in the file on which I'm commenting) please remove the quotes from the property names unless they are necessary.

2. Source code of the form `f: function() {}` and `f: () => {}` must be written in the form `f() {}`.

File: react/features/toolbar/defaultToolbarButtons.js [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+declare var APP: Object;
+declare var JitsiMeetJS: Object;
+
+/**
+ * All toolbar buttons' descriptions.
+ */
+export default {
+    'microphone': {
+        id: 'toolbar_button_mute',
```
Comment (lyubomir) [inline react/features/toolbar/defaultToolbarButtons.js]: I do not approve of the id that is a magic string because it's error prone. Search for the string `'toobar_button_mute'` and you'll find it 4 times. When you combine it with my other comment about properties with values functions and how they are to be written, and you comply with camel case for function names afterwards, I'd like you to think of an elegant way to use functions/function names instead of magic numbers.

File: react/features/toolbar/actions.web.js [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+                toolbarTimeoutObject = null;
+            }
+            toolbarTimeoutObject = setTimeout(() => {
+                dispatch(hideToolbar());
+            }, timeout);
+            toolbarTimeout = interfaceConfig.TOOLBAR_TIMEOUT;
+        }
+
+        // Show/hide desktop sharing button.
+        dispatch(showDesktopSharingButton());
```
Comment (saghul) [inline react/features/toolbar/actions.web.js]: Why is this here?

File: react/features/toolbar/components/Toolbar.native.js [hunk 6]
```diff
@@ -1,41 +1,39 @@
...
 import { MEDIA_TYPE, toggleCameraFacingMode } from '../../base/media';
 import { Container } from '../../base/react';
 import { ColorPalette } from '../../base/styles';
+import { beginRoomLockRequest } from '../../room-lock';
 
-import { AbstractToolbar, _mapStateToProps } from './AbstractToolbar';
+import {
+    _mapDispatchToProps as _abstractMapDispatchToProps,
+    _mapStateToProps as _abstractMapStateToProps
+} from './AbstractToolbar';
```
Comment (saghul) [inline react/features/toolbar/components/Toolbar.native.js]: If there will no longer be an abstract toolbar, this file shouldn't exist and we should use some functions.js

File: react/features/toolbar/components/ToolbarButton.web.js [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+    }
+
+    /**
+     * Wrapper on on click handler props for current button.
+     *
+     * @param {Event} event - Click event object.
+     * @returns {void}
+     * @private
+     */
+    _onClick(event) {
```
Comment (saghul) [inline react/features/toolbar/components/ToolbarButton.web.js]: the previous function has a flow annotation, can you please be more consistent adding them?

File: react/features/toolbar/components/secondaryToolbarHandlers.js (hunk 2 above)
Comment (saghul) [inline react/features/toolbar/components/secondaryToolbarHandlers.js]: Seconded.

File: react/features/toolbar/defaultToolbarButtons.js [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+        enabled: true,
+        shortcut: 'C',
+        shortcutAttr: 'toggleChatPopover',
+        shortcutFunc() {
+            JitsiMeetJS.analytics.sendEvent('shortcut.chat.toggled');
+            APP.UI.toggleChat();
+        },
+        shortcutDescription: 'keyboardShortcuts.toggleChat',
+        sideContainerId: 'chat_container',
+        html: <span className = 'badge-round'>
```
Comment (saghul) [inline react/features/toolbar/defaultToolbarButtons.js]: This looks like it doesn't belong here. I understand you want to group all stuff related to a button somewhere, but this looks weird here.

File: react/features/toolbar/defaultToolbarButtons.js [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+        classNames: [ 'button' ],
+        hidden: true // will be displayed once
+                     // the recording functionality is detected
+    },
+    'sharedvideo': {
+        enabled: true,
+        id: 'toolbar_button_sharedvideo',
+        tooltipKey: 'toolbar.sharedvideo',
+        classNames: [ 'button', 'icon-shared-video' ],
+        popups: [
```
Comment (saghul) [inline react/features/toolbar/defaultToolbarButtons.js]: ditto

File: react/features/toolbar/functions.web.js [hunk 10]
```diff
@@ -0,0 +1,159 @@
+/* @flow */
+
+import SideContainerToggler
+    from '../../../modules/UI/side_pannels/SideContainerToggler';
+
+import defaultToolbarButtons from './defaultToolbarButtons';
+
+declare var AJS: Object;
+declare var interfaceConfig: Object;
+declare var $: Function;
```
Comment (saghul) [inline react/features/toolbar/functions.web.js]: we keep declaring these (and similar globals) on many files. Can you look into putting all those into a file which flow uses automagically? Similarly to how flow-typed does it for other projects.

File: react/features/toolbar/functions.web.js [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+ *
+ * @param {Object} props - Props set to the React component.
+ * @returns {Object}
+ */
+export function getButtonAttributesByProps(props: Object): MapOfAttributes {
+    const { classNames } = props;
+    const classNameCollection = [ ...classNames ];
+
+    if (props.toggled) {
+        classNameCollection.push('toggled');
```
Comment (saghul) [inline react/features/toolbar/functions.web.js]: is "toggled" a term we got from the previous code? I found it a bit surprising.

File: react/features/toolbar/reducer.js [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+ * @returns {Object}
+ * @private
+ */
+function _setButton(state, { key, button }): Object {
+    const {
+        primaryToolbarButtons,
+        secondaryToolbarButtons
+    } = state;
+    let selectedButton = primaryToolbarButtons.get(key);
+    let place = 'primaryToolbarButtons';
```
Comment (saghul) [inline react/features/toolbar/reducer.js]: do we support setting a button in both toolbars at the same time?

File: react/features/toolbar/actions.web.js [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+export * from './actions.native';
+
+declare var $: Function;
+declare var APP: Object;
+declare var config: Object;
+declare var interfaceConfig: Object;
+
+let toolbarTimeout = interfaceConfig.INITIAL_TOOLBAR_TIMEOUT;
+let toolbarTimeoutObject;
+
```
Comment (lyubomir) [inline react/features/toolbar/actions.web.js]: No global variables!

File: react/features/toolbar/functions.web.js (hunk 10 above)
Comment (lyubomir) [inline react/features/toolbar/functions.web.js]: Let's look into it outside of this PR though. And let's do it sooner rather than later.

File: react/features/toolbar/defaultToolbarButtons.js [hunk 14]
```diff
@@ -0,0 +1,334 @@
...
+        },
+        focus: ':input:first'
+    });
+}
+
+/**
+ * All toolbar buttons' descriptions.
+ */
+export default {
+    microphone: {
```
Comment (lyubomir) [inline react/features/toolbar/defaultToolbarButtons.js]: Now that these look like simple object property names, it should be obvious that our coding style requires:
1. jsdocs on everything. I believe at least the top-level objects which describe buttons should have jsdocs.
2. Alphabetical sorting. I don't see any advantage to the random order of button descriptors.

File: react/features/toolbar/defaultToolbarButtons.js [hunk 15]
```diff
@@ -0,0 +1,334 @@
...
+            } else {
+                JitsiMeetJS.analytics.sendEvent('toolbar.audio.muted');
+                APP.UI.emitEvent(UIEvents.AUDIO_MUTED, true, true);
+            }
+        },
+        popups: [
+            {
+                id: 'micMutedPopup',
+                className: 'loginmenu',
+                dataAttr: 'toolbar.micMutedPopup'
```
Comment (lyubomir) [inline react/features/toolbar/defaultToolbarButtons.js]: Sort the object properties in alphabetical order please. i don't see any advantage to this random order. I don't think anyone cares more about `id` than about `className` or `dataAttr`. These are mere collections of key-value pairs.

File: react/features/toolbar/defaultToolbarButtons.js [hunk 16]
```diff
@@ -0,0 +1,334 @@
...
+            JitsiMeetJS.analytics.sendEvent('toolbar.hangup');
+            APP.UI.emitEvent(UIEvents.HANGUP);
+        },
+        tooltipKey: 'toolbar.hangup'
+    },
+    raisehand: {
+        classNames: [ 'button', 'icon-raised-hand' ],
+        content: 'Raise Hand',
+        enabled: true,
+        i18n: 'toolbar.raiseHand',
```
Comment (lyubomir) [inline react/features/toolbar/defaultToolbarButtons.js]: I don't think that `i18n` is effectively used anywhere anymore. I changed an use of `i18n` to `key` in my commit  in order to fix a runtime error. Looking back now, I think I may have broken `i18n` but I see that not every button has `i18n` and your use of it which raised the runtime error was because you relied that `i18n` was always defined. Can you please figure out what's going on with `i18n` and make sure that it's actually eventually used?

File: modules/UI/side_pannels/chat/Chat.js [hunk 17]
```diff
@@ -49,30 +50,41 @@ var CHAT_CONTAINER_ID = "chat_container";
...
-            UIUtil.getTextWidth(unreadMsgElement)) / 2;
-        var topIndent = (UIUtil.getTextHeight(chatButtonElement) -
-            UIUtil.getTextHeight(unreadMsgElement)) / 2 - 5;
+        const leftIndent
+            = (UIUtil.getTextWidth(chatButtonElement)
+                    - UIUtil.getTextWidth(unreadMsgElement))
+                / 2;
+        const topIndent
+            = (UIUtil.getTextHeight(chatButtonElement)
+                        - UIUtil.getTextHeight(unreadMsgElement))
```
Comment (lyubomir) [inline modules/UI/side_pannels/chat/Chat.js]: How is the indentation wrong? It reflects the order in which the arithmetic operators are applied. It's quite a complex expression. Something is reduced by 5 units, that something is the result of dividing something other by 2, and that something other is the difference between heights of the two HTML elements. I applied that formatting and I stand by it.

File: modules/UI/side_pannels/chat/Chat.js [hunk 18]
```diff
@@ -49,30 +50,41 @@ var CHAT_CONTAINER_ID = "chat_container";
...
+                    / 2
+                - 5;
 
         unreadMsgElement.setAttribute(
-            'style',
-                'top:' + topIndent +
-                '; left:' + leftIndent + ';');
+                'style',
```
Comment (lyubomir) [inline modules/UI/side_pannels/chat/Chat.js]: The original was incorrectly indented because the function call is given 2 arguments. I made sure the 2 arguments have the same level of indentation. If you want just 4 spaces of indentation instead of 8, fine, I can take care of it.

File: react/features/conference/components/Conference.native.js (hunk 1 above)
Comment (lyubomir) [inline react/features/conference/components/Conference.native.js]: These are calling dispatch with action creators which may or may not return an asynchronous action so I see no way of saying whether and how much the whole thing is asynchronous or synchronous. I'm OK with the names. If you'd like to apply some other naming scheme, I'd rather we talk it in a meeting and I'll go and apply it outside this PR.

File: react/features/toolbar/actions.native.js [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+export function changeLocalRaiseHand(isRaisedHand: boolean): Function {
+    return (dispatch: Dispatch<*>, getState: Function) => {
+        const buttonName = 'raisehand';
+        const state = getState();
+        const { secondaryToolbarButtons } = state['features/toolbar'];
+        const button = secondaryToolbarButtons.get(buttonName);
+
+        button.toggled = isRaisedHand;
+
+        dispatch(setToolbarButton(buttonName, button));
```
Comment (lyubomir) [inline react/features/toolbar/actions.native.js]: I don't like this naming because it raises unnecessary complexity and that's why there's a rule in our coding style that 1 abstraction should be given 1 name. I cannot understand why you call it `buttonName` here and `key` in `setToolbarButton`. So when I go and try to understand how the whole thing works I have to keep a map of aliases for one and the same abstraction.

File: react/features/film-strip/components/FilmStrip.js [hunk 20]
```diff
@@ -112,7 +113,15 @@ function _mapStateToProps(state) {
...
-        _participants: state['features/base/participants']
+        _participants: state['features/base/participants'],
+
+        /**
+         * The indicator which determines whether toolbar is visible.
+         *
+         * @private
+         * @type {boolean}
+         */
+        _visible: !state['features/toolbar'].visible
```
Comment (lyubomir) [inline react/features/film-strip/components/FilmStrip.js]: Because the React Component FilmStrip is used on mobile only at the time of this writing and on mobile the film strip is visible when the toolbar is not.

File: react/features/toolbar/actions.web.js [hunk 21]
```diff
@@ -0,0 +1,307 @@
...
+        }
+
+        if (!visible) {
+            dispatch(setToolbarVisibility(true));
+            dispatch(setSubjectSlideIn(true));
+
+            updateTimeout = true;
+        }
+
+        if (updateTimeout) {
```
Comment (lyubomir) [inline react/features/toolbar/actions.web.js]: The variable `updateTimeout` seems superficial to me. The condition of `if (updateTimeout) {` will be true only if the condition of `if (!visible) {` is true. So why are there 2 separate `if`s instead of 1 then?

File: modules/UI/side_pannels/chat/Chat.js (hunk 17 above)
Comment (saghul) [inline modules/UI/side_pannels/chat/Chat.js]: Hum, maybe it's the rendering on GH, but it looks like the line starting with "-" is 2x indented than the one starting with "=" is that intentional?

File: modules/UI/side_pannels/chat/Chat.js (hunk 18 above)
Comment (saghul) [inline modules/UI/side_pannels/chat/Chat.js]: Yeah, I think 4 would be nicer.

File: react/features/conference/components/Conference.native.js (hunk 1 above)
Comment (saghul) [inline react/features/conference/components/Conference.native.js]: Ok, let's discuss it on more general terms. As a general rule, I usually use onXXX for functions that are passed around and called by some other entity when some event happens, not by directly calling them myself.

In this case, onConnect suggests me that the connection was made, but it actually wasn't, the component was mounted and we want to connect after that.

File: react/features/film-strip/components/FilmStrip.js (hunk 20 above)
Comment (saghul) [inline react/features/film-strip/components/FilmStrip.js]: Ah true that! I was looking at the change in isolation and forgot that. A comment might help other readers.

File: react/features/toolbar/functions.web.js (hunk 10 above)
Comment (saghul) [inline react/features/toolbar/functions.web.js]: Agreed!

File: react/features/film-strip/components/FilmStrip.js (hunk 20 above)
Comment (BeatC) [inline react/features/film-strip/components/FilmStrip.js]: Sure, let's add some explanation. It looks confusing if you see this code the first time.

File: react/features/toolbar/reducer.js (hunk 12 above)
Comment (BeatC) [inline react/features/toolbar/reducer.js]: You should dispatch two separate actions to do that. And yes, for example we can hide `desktopsharing` button from the primary toolbar and `sip` from secondary one.

File: react/features/toolbar/defaultToolbarButtons.js (hunk 4 above)
Comment (BeatC) [inline react/features/toolbar/defaultToolbarButtons.js]: I left `id` property in these objects because we use id to set button tooltips. But actual differentiation between buttons is now implemented via their names in `interface_config`

File: react/features/toolbar/actions.web.js [hunk 22]
```diff
@@ -0,0 +1,307 @@
...
+}
+
+/**
+ * Action that reset always visible toolbar to default state.
+ *
+ * @returns {Function}
+ */
+export function resetAlwaysVisibleToolbar(): Function {
+    return (dispatch: Dispatch<*>) => {
+        const alwaysVisible = config.alwaysVisibleToolbar === true;
```
Comment (BeatC) [inline react/features/toolbar/actions.web.js]: I have no idea whether this config option is used. But I ported it from the old app.

Comment (saghul) [inline react/features/toolbar/actions.web.js]: @yanas can you chime in?

File: react/features/toolbar/actions.native.js (hunk 19 above)
Comment (BeatC) [inline react/features/toolbar/actions.native.js]: Changed `key` names to `buttonName`s

File: react/features/toolbar/actions.web.js (hunk 21 above)
Comment (BeatC) [inline react/features/toolbar/actions.web.js]: Merged two `if`s

File: react/features/toolbar/components/ToolbarButton.web.js (hunk 7 above)
Comment (BeatC) [inline react/features/toolbar/components/ToolbarButton.web.js]: Done!

File: react/features/toolbar/actions.web.js [hunk 23]
```diff
@@ -0,0 +1,307 @@
...
+ * @returns {Function}
+ */
+export function showSharedVideoButton(): Function {
+    return (dispatch: Dispatch<*>) => {
+        const buttonId = 'toolbar_button_sharedvideo';
+        const shouldShow
+            = UIUtil.isButtonEnabled('sharedvideo')
+                && !config.disableThirdPartyRequests;
+
+        if (shouldShow) {
```
Comment (BeatC) [inline react/features/toolbar/actions.web.js]: Yep, in this case !shouldShow will always be equal to `false` but let's replace literal by expression.

File: react/features/toolbar/actions.web.js [hunk 24]
```diff
@@ -0,0 +1,307 @@
...
+        const alwaysVisible = config.alwaysVisibleToolbar === true;
+
+        dispatch(setAlwaysVisibleToolbar(alwaysVisible));
+    };
+}
+
+/**
+ * Signals that unclickable property of profile button should change its value.
+ *
+ * @param {boolean} unclickable - Shows whether button is unclickable.
```
Comment (BeatC) [inline react/features/toolbar/actions.web.js]: No, it is not disabled state. Because if button is disabled we're still showing tooltips for them. Also styling for disabled and unclickable buttons differs. At this moment we're only setting profile button in unclickable state.

File: react/features/toolbar/actions.web.js [hunk 25]
```diff
@@ -0,0 +1,307 @@
...
+        } = state['features/toolbar'];
+
+        if (alwaysVisible) {
+            return;
+        }
+
+        clearTimeout(toolbarTimeoutObject);
+        toolbarTimeoutObject = null;
+
+        if (force !== true
```
Comment (BeatC) [inline react/features/toolbar/actions.web.js]: Sure, overlooked while being simplified the logic from the old app.

File: react/features/toolbar/actionTypes.js [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * {
+ *     type: SET_SUBJECT_SLIDE_IN,
+ *     subjectSlideIn: boolean
+ * }
+ */
+export const SET_SUBJECT_SLIDE_IN = Symbol('SET_SUBJECT_SLIDE_IN');
+
+/**
+ * The type of the action which signals that a value for
```
Comment (BeatC) [inline react/features/toolbar/actionTypes.js]: Done!

File: webpack.config.js [hunk 27]
```diff
@@ -56,6 +56,7 @@ if (minimize) {
 // jitsi-meet such as app.bundle.js and external_api.js.
 var config = {
     devServer: {
+        host: '0.0.0.0',
```
Comment (lyubomir) [inline webpack.config.js]: I was working with webpack-dev-server on this PR without any problems from my first review . That's why I don't understand why `host` is suddenly necessary. As there's no source code comment explaining the need, I'll just remove it from this PR.

**Comment by yanas** [comment]
Setting a subject is currently broken with this PR. It's not a very popular feature and also the css also seems to be a little bit broken on meet.jit.si currently, but still I see a regression from the Meet behaviour.

**Comment by lyubomir** [comment]
@yana reported the following regression as well:
1. She types a full URL with a room name in the browser's location.
2. The toolbar(s) are rendered.
3. The toolbar(s) disappear.
4. Her local video is displayed.
5. The toolbar(s) are rendered again.

File: react/features/toolbar/actions.web.js [hunk 28]
```diff
@@ -0,0 +1,291 @@
...
+
+        if (isDock) {
+            // First make sure the toolbar is shown.
+            visible || dispatch(showToolbar());
+
+            dispatch(clearToolbarTimeout());
+        } else if (visible) {
+            dispatch(setToolbarTimeout(() => {
+                dispatch(hideToolbar());
+            }, toolbarTimeout));
```
Comment (lyubomir) [inline react/features/toolbar/actions.web.js]: @BeatC, the above is not as simple as possible. Please make an effort in the future to follow my lead. The above is to be written simply as follows: `() => dispatch(hideToolbar())`

File: react/features/toolbar/actions.native.js [hunk 29]
```diff
@@ -0,0 +1,263 @@
...
+/**
+ * Signals that toolbar visibility value should be changed.
+ *
+ * @param {boolean} visible - Value to be set in redux store.
+ * @returns {{
+ *     type: SET_TOOLBAR_VISIBILITY,
+ *     visible: boolean
+ * }}
+ */
+export function setToolbarVisibility(visible: boolean): Object {
```
Comment (lyubomir) [inline react/features/toolbar/actions.native.js]: @BeatC, the names `setToolbarVisibility`, `_onToolbarVisibilityChange`, and `isVisible` do not comply with our coding style which requires 1 name per abstraction. Since the property is called `visible`, a respective function argument (e.g. `_onToolbarVisibilityChange`) should be called `visible` as well, not `isVisible`. Respectively, the function should be called `setToolbarVisible` and, generally, `_onToolbarVisibleChanged` (but in your case the whole name `_onToolbarVisibilityChange` is inappropriate because it's not a handler which responds to changes in the value of a property `visible` of a `Toolbar`). I've taken care of some of the violations of the coding style but you're really expect to be taking care of these (especially since I've repeated the rule in question to you multiple times).

File: react/features/toolbar/actions.web.js [hunk 30]
```diff
@@ -0,0 +1,291 @@
...
+ */
+export function showToolbar(timeout: number = 0): Object {
+    return (dispatch: Dispatch<*>, getState: Function) => {
+        if (interfaceConfig.filmStripOnly) {
+            return;
+        }
+
+        const state = getState();
+        const { toolbarTimeout, visible } = state['features/toolbar'];
+        const finalTimeout = timeout ? timeout : toolbarTimeout;
```
Comment (lyubomir) [inline react/features/toolbar/actions.web.js]: The above is not the simplest way to write the source code. Please pay attention and take care of such cases yourself in the future: `timeout ? timeout : toolbarTimeout` is equivalent to `timeout || toolbarTimeout`.

File: react/features/toolbar/actions.web.js [hunk 31]
```diff
@@ -0,0 +1,291 @@
...
+
+        if (alwaysVisible) {
+            return;
+        }
+
+        dispatch(clearToolbarTimeout());
+
+        if (!force && (isHovered
+                || APP.UI.isRingOverlayVisible()
+                || SideContainerToggler.isVisible())) {
```
Comment (lyubomir) [inline react/features/toolbar/actions.web.js]: @BeatC, the formatting of the above does not reflect the order in which the boolean expressions will be evaluated. Please comply with our coding style (e.g. the consistency rule) and make sure to properly format the source code you submit as ready for PR review.

File: react/features/toolbar/reducer.js [hunk 32]
```diff
@@ -0,0 +1,137 @@
...
+
+    return {
+        ...getDefaultToolbarButtons(),
+
+        alwaysVisible: false,
+        isHovered: false,
+        subjectSlideIn: false,
+        timeoutId: null,
+        toolbarTimeout,
+        visible: false
```
Comment (lyubomir) [inline react/features/toolbar/reducer.js]: 1. Seriously, @BeatC, when I pointed out to you in this PR that our coding style requires jsdocs everywhere, I really meant that. You're defining properties above.

2. I find it inconsistent to use `boolean` variable naming with and without `is`. Please stick to one style. As we have `visible`, please rename `isHovered`, `isDock`, etc. to be consistent with `visible`.

File: react/features/toolbar/components/PrimaryToolbar.web.js [hunk 33]
```diff
@@ -0,0 +1,294 @@
...
+ *     _toolbarButtons: Map,
+ *     _videoMuted: boolean
+ *}}
+ * @private
+ */
+function _mapStateToProps(state: Object): Object {
+    const { primaryToolbarButtons } = state['features/toolbar'];
+
+    return {
+        ...abstractMapStateToProps(state),
```
Comment (yanas) [inline react/features/toolbar/components/PrimaryToolbar.web.js]: abstractMapStateToProps properties don't seem to be used in this class or am I missing something?

**Comment by yanas** [comment]
Just a quick note that most of the issues I have reported to Lyubo and here in the ticket have been fixed. The only thing I still see is the subject disappearing. It can be added in a separate PR.

**Comment by yanas** [comment]
Tested the last version of this PR and the subject is now fixed as well.

File: react/features/toolbar/actions.web.js (hunk 22 above)
Comment (virtuacoplenny) [inline react/features/toolbar/actions.web.js]: alwaysVisibleToolbar looks to be used in the jitsi-torture tests. I'm betting it's used to ensure the toolbar is clickable without worrying about hiding/showing.
